config/fcos/v1_5_exp: use array of extension instead of string (#316)
coreos/butane: docs/config-fcos-v1_5-exp.md

    * **_threshold_** (int): sets the minimum number of pieces required to decrypt the device. Default is 1.
  * **_mirror_** (object): describes mirroring of the boot disk for fault tolerance.
    * **_devices_** (list of strings): the list of whole-disk devices (not partitions) to include in the disk array, referenced by their absolute path. At least two devices must be specified.

> * **_extensions_** (list of objects): the list of packages to layer on top of the OS.

Reviewer: Are these actually arbitrary package names, or are they a more restricted set of strings?

Author: anything on the repos or rpm-ostree install allows

Reviewer: Hmm, maybe I'm missing something here.  The existing MachineConfig `extensions` field supports only a restricted set of extension names, not arbitrary packages, and I had understood that this field was intended to be similar.  If this is intended for package installation, should we be calling it `packages` instead of `extensions`?

Reviewer: ^ cc @jlebon

Author: It's similar as in it allows you to extend your OS. But this effort is part of the layering enhancement, as part of that we are trying to enable users to layer whatever they need. We are still having multiple conversation about the enhancement so naming could change. The idea of using Butane to declare the extensions/packages allow us to do something like: https://github.com/coreos/fcos-derivation-example/blob/main/Dockerfile but in a way that we can add more validation around it. A proof of concept of the effort is here: https://github.com/mkenigs/coreos-derive/tree/main/prototype#readme

Reviewer 2: For the MCO the fact that the available packages/extensions are lifecycle bound and tested with the cluster is an important aspect of things, IMO. FWIW this came from https://github.com/openshift/enhancements/blob/master/enhancements/rhcos/extensions.md

I can definitely see the argument for calling this `packages:` though.

(But I also have been trying for years now to create a distinction between "base image" and "addons/extensions", and so with this it's much more clear that one can't just do e.g. `podman-3.2.0` to override something from the base OS)

There's also the "why not both" solution of
```
extensions:
  packages:
    - name: usbguard
```

I don't feel *really* strongly on this TBH.

Reviewer 3: The allowed set from the rpm-ostree side is really any "provides" which include e.g. package names, virtual provides, and even filepaths. Not sure if there's a good reason to try to be stricter here than what rpm-ostree supports.

I like the connotation of extensions to mean packages which really have to be on the host, and also like that we're leaving the door open for other kinds of extensions, so personally would leave it as is.

Minor suggestion, maybe just

```suggestion
* **_extensions_** (list of objects): the list of additional packages to be installed.
```

to be less implementation-y?

>   * **name** (string): the name of the package.

Reviewer: This is correctly marked required here, but we're not actually enforcing that.  It's not _strictly_ necessary to do so (and it's unlikely that a user will forget the field until we have other fields in the struct) but would provide better UX than failing at runtime.  Could we add validation in `validate.go`, and a test for the validation?
